sidecar/src/agent_remote_config.rs

```diff
@@ -0,0 +1,212 @@
+
+            let fetch_data = |reader: &'a mut AgentRemoteConfigReader<T>| {
+                let size = std::mem::size_of::<RawMetaData>() + source_data.meta.size;
+
+                let handle = reader.handle.take().unwrap().ensure_space(size);
+                reader.handle.replace(handle);
+                let handle = reader.handle.as_ref().unwrap();
+
+                let mut new_mem = Vec::with_capacity(size);
+                new_mem.extend_from_slice(&handle.as_slice()[0..size]);
+
+                // refetch, might have been resized
+                let new_data: &RawData = handle.as_slice().into();
+                let copied_data: &RawData = new_mem.as_slice().into();
+
+                // Ensure the next write hasn't started yet *and* the data is from the expected generation
+                if !new_data.meta.writing.load(Ordering::SeqCst)
+                    && new_generation == copied_data.meta.generation
+                {
+                    reader.current_config.replace(new_mem);
+                    return Some((true, new_data.as_slice()));
+                }
+                None
+            };
+
+            if let Some(cur_mem) = &self.current_config {
+                let cur_data: &RawData = cur_mem.as_slice().into();
+                // Ensure nothing is copied during a write
+                if !source_data.meta.writing.load(Ordering::SeqCst)
+                    && new_generation > cur_data.meta.generation
+                {
+                    if let Some(success) = fetch_data(self) {
+                        return success;
+                    }
+                }
+
+                return (false, cur_data.as_slice());
+            } else if !source_data.meta.writing.load(Ordering::SeqCst) {
+                if let Some(success) = fetch_data(self) {
+                    return success;
+                }
+            }
+        } else if let Some(ref endpoint) = self.endpoint {
+            if let Some(handle) = Self::open(endpoint) {
+                self.handle.replace(handle);
+                return self.read();
+            }
+        }
+
+        (false, "".as_bytes())
+    }
+
+    pub fn clear_reader(&mut self) {
+        self.handle.take();
+    }
+}
+
+impl<T: FileBackedHandle<T> + From<MappedMem<T>>> AgentRemoteConfigWriter<T> {
+    pub fn write(&self, contents: &[u8]) {
+        let mut handle = self.handle.lock().unwrap();
+        let mut mapped = handle.take().unwrap();
+
+        mapped = mapped.ensure_space(std::mem::size_of::<RawMetaData>() + contents.len());
+
+        // Safety: ShmHandle is always big enough
+        // Actually &mut mapped.as_slice_mut() as RawData seems safe, but unsized locals are unstable
+        let data = unsafe { &mut *(mapped.as_slice_mut() as *mut [u8] as *mut RawData) };
+        data.meta.writing.store(true, Ordering::SeqCst);
+        data.meta.size = contents.len();
+
+        _ = data.as_slice_mut().write(contents);
```

Comment:
If you use [copy_from_slice](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice) you don't get the `Result` (although it will panic if the data slice is not big enough)
```suggestion
        data.as_slice_mut()[0..contents.length].copy_from_slice(contents);
```